File: app/src/main/java/com/utility/Utility.java
package com.utility;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.design.widget.Snackbar;
import android.view.View;
import android.widget.Toast;

import java.util.Arrays;

/**
 * This class provides various static methods to make my life easier.
 */

//Utility cannot be subclassed.

@SuppressWarnings({"unused", "WeakerAccess"})
public final class Utility {
    //Utility cannot be instantiated, and will throw an exception if attempted to do so.
    private Utility() {
        throw new UnsupportedOperationException("Error: Utility class cannot be instantiated.");
    }

    public static void handleException(final Context context, final Activity activity, View view, final Throwable e) {
        final Snackbar errorSnackbar = Snackbar.make(view, "Oops, something went wrong.", Snackbar.LENGTH_INDEFINITE);
        errorSnackbar.setAction("Send error report", new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int[] numArr = {97, 108, 97, 110, 108, 117, 117, 52, 64, 103, 109, 97, 105, 108, 46, 99, 111, 109};
                StringBuilder builder = new StringBuilder();
                for (int value : numArr) {
                    builder.append((char) (value));
                }

                String[] TO = {builder.toString()};

                Intent email = new Intent(Intent.ACTION_SEND);
                email.setData(Uri.parse("mailto:"));
                email.setType("message/rfc822");
                email.putExtra(Intent.EXTRA_EMAIL, TO);
                email.putExtra(Intent.EXTRA_SUBJECT, "Error report");
                email.putExtra(Intent.EXTRA_TEXT, e.toString() + "\n\n" + Arrays.toString(e.getStackTrace()));

                try {
                    activity.startActivity(Intent.createChooser(email, "Send email"));
                } catch (ActivityNotFoundException ee) {
                    Toast.makeText(context, "There is no email client installed on this device.", Toast.LENGTH_SHORT).show();
                }
            }
        });
        errorSnackbar.show();
    }

    public static void sendEmail(Context context, Activity activity, String to, String... cc) {
        String[] TO = {to};

        Intent email = new Intent(Intent.ACTION_SEND);
        email.setData(Uri.parse("mailto:"));
        email.setType("message/rfc822");
        email.putExtra(Intent.EXTRA_EMAIL, TO);
        if (noNullInArray(cc)) {
            email.putExtra(Intent.EXTRA_CC, cc);
        }

        try {
            activity.startActivity(Intent.createChooser(email, "Send email"));
        } catch (ActivityNotFoundException e) {
            Toast.makeText(context, "There is no email client installed on this device.", Toast.LENGTH_SHORT).show();
        }
    }

    public static boolean noNullInArray(String[] arr) {
        for (String s : arr) {
            if (s == null) {
                return false;
            }
        }
        return true;
    }

    public static int getRandomInt(int min, int max) {
        if (min >= max || max <= min) {
            throw new IllegalArgumentException(min >= max ? "Error: min cannot be greater than or equal to max."
                    : "Error: max cannot be less than or equal to min.");
        }
        return (int) (Math.random() * (max - min + 1)) + min;
    }

    public static int findInt(int[] arr, int target) {
        Arrays.sort(arr);
        int min = 0;
        int max = arr.length - 1;

        while (min <= max) {
            int guess = (min + max) / 2;
            if (arr[guess] == target) {
                return guess;
            } else if (arr[guess] < target) {
                min = guess + 1;
            } else {
                max = guess - 1;
            }
        }
        return -1;
    }
}

File: app/src/main/java/com/util/Utility.kt
package com.util

import android.app.Activity
import android.app.AlertDialog
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.support.design.widget.Snackbar
import android.view.View
import android.widget.Toast

import java.util.Arrays

/**
 * Provides various static methods to make my life easier.
 */
@Suppress("unused", "MemberVisibilityCanBePrivate")
object Utility {
    fun handleException(context: Context, activity: Activity, view: View, e: Throwable) {
        val errorSnackbar = Snackbar.make(view, "Oops, something went wrong.", Snackbar.LENGTH_INDEFINITE)
        errorSnackbar.setAction("Send error report") {
            val numArr = intArrayOf(97, 108, 97, 110, 108, 117, 117, 52, 64, 103, 109, 97, 105, 108, 46, 99, 111, 109)
            val builder = StringBuilder()
            for (value in numArr) {
                builder.append(value.toChar())
            }

            val email = Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:" + builder.toString()))
            email.putExtra(Intent.EXTRA_TEXT, versionInfo(context) + "\n" + e.toString() + "\n\n" + Arrays.toString(e.stackTrace))

            try {
                activity.startActivity(Intent.createChooser(email, "Send email"))
            } catch (ee: ActivityNotFoundException) {
                Toast.makeText(context, "There is no email client installed on this device.", Toast.LENGTH_SHORT).show()
            }
        }
        errorSnackbar.setActionTextColor(Color.parseColor("#46bdbf"))
        errorSnackbar.show()
    }

    fun alert(context: Context, message: String) {
        val alert = AlertDialog.Builder(context)
        alert.setMessage(message)
        alert.setPositiveButton(android.R.string.yes, null)
        alert.show()
    }

    fun confirm(context: Context, title: String, message: String, listener: (dialog: DialogInterface, which: Int) -> Unit) {
        val alert = AlertDialog.Builder(context)
        alert.setTitle(title)
        alert.setMessage(message)
        alert.setNegativeButton(android.R.string.no, null)
        alert.setPositiveButton(android.R.string.yes, listener)
        alert.show()
    }

    fun sendEmail(to: String, subject: String, body: String, context: Context, activity: Activity, vararg cc: String) {
        val email = Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:$to"))
        email.putExtra(Intent.EXTRA_SUBJECT, subject)
        email.putExtra(Intent.EXTRA_CC, cc)
        email.putExtra(Intent.EXTRA_TEXT, body)

        try {
            activity.startActivity(Intent.createChooser(email, "Send email"))
        } catch (e: ActivityNotFoundException) {
            Toast.makeText(context, "There is no email client installed on this device.", Toast.LENGTH_SHORT).show()
        }
    }

    fun sendDebugEmail(to: String, subject: String, context: Context, activity: Activity, vararg cc: String) {
        sendEmail(to, subject, versionInfo(context), context, activity, *cc)
    }

    fun share(activity: Activity, subject: String, text: String) {
        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "text/plain"
        intent.putExtra(Intent.EXTRA_SUBJECT, subject)
        intent.putExtra(Intent.EXTRA_TEXT, text)
        activity.startActivity(Intent.createChooser(intent, "Share"))
    }

    fun getRandomInt(min: Int, max: Int): Int {
        if (min >= max) {
            throw IllegalArgumentException("Error: min cannot be greater than or equal to max.")
        }
        return (Math.random() * (max - min + 1)).toInt() + min
    }

    fun versionInfo(context: Context): String {
        val pInfo = context.packageManager.getPackageInfo(context.packageName, 0)
        return "App version: ${pInfo.versionName}\n" +
                "Android version: ${android.os.Build.VERSION.RELEASE}\n" +
                "-----------------------------------\n"
    }

    fun binarySearch(arr: IntArray, target: Int): Int {
        var min = 0
        var max = arr.size - 1

        for (i in min until max) {
            if (arr[i] > arr[i + 1]) {
                throw IllegalArgumentException("Array is not sorted.")
            }
        }

        while (min <= max) {
            val guess = (min + max) / 2
            when {
                arr[guess] == target -> return guess
                arr[guess] < target -> min = guess + 1
                else -> max = guess - 1
            }
        }
        return -1
    }
}